data_analysis/correction_framework.py

```diff
@@ -0,0 +1,247 @@
+                data[3].float(),
+                data[4].float,
+                data[5],
+            )
+            if (
+                m_rec.local_time < self.sunset
+            ):  # unsure if this will work or if a special time comp function is needed
+                n_file.readline()
+                continue
+            self.min_records.append(m_rec)
+            transtr = n_file.readline()
+        n_file.close()
+
+        m_file = open(morning_file)
+        transtr = m_file.readline()
+        while len(transtr) != 0:
+            data = transtr.split(";")
+            data[0] = data[0].replace("T", " ")
+            data[1] = data[1].replace("T", " ")
+            m_rec = minute_record(
+                Time(data[0]),
+                Time(data[1]),
+                data[2].float(),
+                data[3].float(),
+                data[4].float,
+                data[5],
+            )
+            if (
+                m_rec.local_time > self.sunrise
+            ):  # unsure if this will work or if a special time comp function is needed
+                m_file.readline()
+                break
+            self.min_records.append(m_rec)
+            transtr = m_file.readline()
+        m_file.close()
+
+        self.valid_for_use = True
+        self.exclusion_reason = "None"
+
+    # mark data as invalid based on various night conditions
+    def grade_data(self):
+        # taking the average night to be ten hours (600 minutes), if more than three hours (180 minutes) are missing,
+        # invalidate the data based on insufficient records
+        if len(self.min_records) < 420:
+            self.valid_for_use = False
+            self.exclusion_reason = "Insufficient Records for Night"
+
+        # if the moon would have been obscured by bad weather for more then three hours, invalidate based on weather conditions
+        # use something other than python-weather?
+
+    # given a function on a float, updates all temperature, frequency, or voltage values
+    def apply_correction(self, field_to_update: str, funct):
+        if field_to_update == "TEMP":
+            for rec in self.min_records:
+                rec.temperature = funct(rec.temperature)
+        if field_to_update == "FREQ":
+            for rec in self.min_records:
+                rec.frequency = funct(rec.frequency)
+        if field_to_update == "VOLT":
+            for rec in self.min_records:
+                rec.voltage = funct(rec.voltage)
+        else:
+            print("error: field not recognized/field should not be updated")
+
+    # export night record to a new file
+    def rec_export(self, filename: str):
+        new_file = open(filename, "x")
+        if self.valid_for_use == False:
+            new_file.write("INVALID\n")
+            new_file.write(self.exclusion_reason)
+        else:
+            # night date -- 2023-01-23 06:00:00.992
+            new_file.write(self.night_date.strptime("%Y-%m-%d %H:%M:%S.%f"))
```

Comment:
This line should be a constant. Set it at the top of the file with something like this, `PROJECT_DATE_FORMAT = "%Y-%m-%d %H:%M:%S.%f"`